### DISCUSSION TITLE
Некорректная работа Filter.removeArtists?

### DISCUSSION THREAD

**Discussion by ilyanikolaich** [Chimildic/goofy#68] [Помощь с ошибкой] [answered]
если я всё правильно понял из описания функции `Filter.removeArtists();`, то код:
```javascript
let Likes = Source.getSavedAlbumTracks();
Filter.dedupArtists(Likes);
let Artists = Source.getArtists({ followed_include: true, });
Filter.removeArtists(Artists,Likes); 
let MyTracks = Source.getArtistsTopTracks(Artists, false);
MyTracks.forEach(tracks => { tracks.splice(1,9); } );
MyTracks = MyTracks.flat(1);
```
должен вернуть массив треков, состоящий из 1 топ трека каждого исполнителя, на которого я подписан, но среди треков которого у меня ни один не лайкнут. Однако это не корректно - среди полученных исполнителей после отработки функции removeArtists - очень много тех, у которых есть мои лайки (причем по многу). И даже если `Filter.removeArtists(MyTracks,Likes);` перенести за получение плейлиста треков и применить уже непосредственно к плейлисту треков - результат точно такой же.
Мне кажется что-то не так, или я что-то не понимаю в описании функции.

**Comment by Chimildic** [comment, marked as answer]
Вы хотите удалить лайки исполнителей или лайки треков? `Filter.removeArtists` удалит исполнителей, которых увидит в `Likes`.

**Reply by ilyanikolaich** [reply]
Да, я хочу, чтобы из `Artists` или из `MyTracks` удалились все исполнители, которые есть в массиве треков `Likes`. Но `Filter.removeArtists` и не удаляет их - ни из массива исполнителей, ни из массива треков.

**Reply by Chimildic** [reply]
Вы уверены в наполнении `Likes`? Туда попадут треки из лайкнутных альбомов `getSavedAlbumTracks`.
Мой дебаг показывает 144 исполнителя до фильтра и 121 после.

**Reply by ilyanikolaich** [reply]
прошу прощения, я - бесстыдная лошара, понадеялся безоглядно на автоподстановку названий функций в редакторе кода, и элементарно не удосужился внимательно посмотреть на то, что получилось.
вместо `Source.getSavedAlbumTracks();` конечно же должно быть **`Source.getSavedTracks();`**

То есть, мне нужен был плейлист, в котором по одной какой-то (пусть самой топовой) композиции от исполнителя, на которого я подписан, но ни одного его трека у нет в моих "лайках".

Это в корне меняет ситуацию, и функция работает _практически_ как нужно, но - благодаря этому все равно выявились определенные фичи, а именно:
конкретно факты:
у меня в "лайках" трек **Hypnotized** исполнения **Purple Disco Machine** (первые в списке) совместно с **Sophie and the Giants**. Оба исполнителя у меня в "подписке"/"лайкнуты". И эти исполнители не отфильтровались `Filter.removeArtists`, - (хотя они оба у меня в "подписке").
Точно так же получилось с еще двумя треками из итогового плейлиста на 183 трека. Более глубоко не анализировал - пока нет времени, но визуально вроде все похоже на правду.
Безусловно, это можно считать "фичей", поскольку здесь всего 1,6% погрешности. Но, принципа ради, было бы интересно понять, почему всё-таки так получилось.

**Reply by Chimildic** [reply]
`removeArtists` смотрит только на первого исполнителя трека. Обычно он основной. 
Можно дополнительно поставить фильтр на удаление треков `removeTracks`, чтобы явно убрать лайки.

Другой вариант, создать массив со всеми исполнителями `savedArtists`. При этом не делая `dedupArtists`.
```js
let savedTracks = Source.getSavedTracks();
let savedArtists = savedTracks.reduce((artists, track) => Combiner.push(artists, track.artists), []);
let followedArtists = Source.getArtists({ followed_include: true, });
Filter.removeArtists(followedArtists, savedArtists);
```

**Reply by ilyanikolaich** [reply]
>> removeArtists смотрит только на первого исполнителя трека.

да, это описано в документации, но - я как раз привел пример, когда она не убирает этого самого первого исполнителя, так как в лайкнутом треке - два исполнителя - и оба они - лайкнуты.

тут больше вопрос не в том, чтобы немедленно получить результат, а в том, чтобы понять, что является причиной такого выполнения.

>> Другой вариант, создать массив со всеми исполнителями savedArtists.

интересное решение, записал!
